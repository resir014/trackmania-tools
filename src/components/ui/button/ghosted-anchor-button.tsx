import * as React from 'react';

import clsx from 'clsx';
import {
  AnchorButtonProps,
  buttonBlockStyles,
  buttonRoundedStyles,
  buttonSizes,
  disabledStyles,
  renderButtonIcon,
} from './utils';

/**
 * A ghosted button with no background when not hovered.
 */
export const GhostedAnchorButton = React.forwardRef<HTMLAnchorElement, AnchorButtonProps>(
  (
    {
      className,
      style,
      type,
      block,
      size = 'md',
      rounded,
      icon,
      iconPosition = 'left',
      iconOnly,
      children,
      ...rest
    },
    ref
  ) => (
    <a
      className={clsx(
        buttonBlockStyles(block, iconPosition),
        buttonSizes(size),
        buttonRoundedStyles(rounded, size),
        'items-center justify-center border border-transparent font-medium text-blue-200 hover:bg-blue-900 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500',
        disabledStyles,
        className
      )}
      ref={ref}
      style={style}
      type={type ?? 'button'}
      {...rest}
    >
      {renderButtonIcon({ icon, size, iconPosition })}
      <span {...(iconOnly ? { className: 'sr-only' } : {})}>{children}</span>
    </a>
  )
);

GhostedAnchorButton.displayName = 'GhostedAnchorButton';
