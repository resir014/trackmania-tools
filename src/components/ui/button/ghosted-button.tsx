import * as React from 'react';

import clsx from 'clsx';
import { Spinner } from '../spinner';
import {
  buttonBlockStyles,
  ButtonProps,
  buttonRoundedStyles,
  buttonSizes,
  disabledStyles,
  renderButtonIcon,
} from './utils';

/**
 * A ghosted button with no background when not hovered.
 */
export const GhostedButton = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      className,
      style,
      type,
      block,
      size = 'md',
      rounded,
      icon,
      iconPosition = 'left',
      iconOnly,
      children,
      isLoading,
      loadingText = 'Loading...',
      disabled,
      ...rest
    },
    ref
  ) => (
    <button
      className={clsx(
        buttonBlockStyles(block, iconPosition),
        buttonSizes(size),
        buttonRoundedStyles(rounded, size),
        'items-center justify-center border border-transparent font-medium text-blue-200 hover:bg-blue-900 focus:outline-none focus:ring-2 focus:ring-green-500 focus:ring-offset-2 ring-offset-gray-900',
        disabledStyles,
        className
      )}
      disabled={isLoading ?? disabled}
      ref={ref}
      style={style}
      type={type ?? 'button'}
      {...rest}
    >
      {renderButtonIcon({
        icon: isLoading ? Spinner : icon,
        size,
        additionalClasses: isLoading ? 'animate-spin' : undefined,
        iconPosition,
        iconOnly,
      })}
      <span {...(iconOnly ? { className: 'sr-only' } : {})}>
        {isLoading ? loadingText : children}
      </span>
    </button>
  )
);

GhostedButton.displayName = 'GhostedButton';
