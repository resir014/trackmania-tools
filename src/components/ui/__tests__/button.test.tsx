/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import React from 'react';

import { PlusIcon } from '@heroicons/react/24/outline';
import { render } from '@testing-library/react';
import { OutlineButton, PrimaryButton, SecondaryButton } from '../button';

describe('Button', () => {
  describe('PrimaryButton', () => {
    test('renders correctly', () => {
      const { container } = render(<PrimaryButton>test button</PrimaryButton>);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('renders with correct colors', () => {
      const { container } = render(<PrimaryButton color="green">test button</PrimaryButton>);

      expect(container.firstChild).toHaveClass('bg-green-600', 'hover:bg-green-700');
    });

    test('renders proper rounded styles', () => {
      const { container } = render(<PrimaryButton rounded>test button</PrimaryButton>);

      expect(container.firstChild).toHaveClass('rounded-full');
    });

    test('renders with correct sizes', () => {
      const { container } = render(<PrimaryButton size="sm">test button</PrimaryButton>);

      expect(container.firstChild).toHaveClass('px-3', 'py-2', 'text-sm', 'leading-4');
    });

    test('renders right icon with the correct classes', () => {
      const { container } = render(
        <PrimaryButton icon={PlusIcon} iconPosition="right">
          test button
        </PrimaryButton>
      );

      expect(container.firstChild).toHaveClass('inline-flex', 'flex-row-reverse');

      expect(container.querySelector('svg')).toHaveClass('ml-2', '-mr-1', 'h-5', 'w-5');
    });

    test('renders custom styles', () => {
      const { container } = render(
        <PrimaryButton className="text-white bg-green-500 focus:ring-green-500" color="none">
          test button
        </PrimaryButton>
      );

      expect(container.firstChild).toHaveClass(
        'text-white',
        'bg-green-500',
        'focus:ring-green-500'
      );
    });
  });

  describe('SecondaryButton', () => {
    test('renders correctly', () => {
      const { container } = render(<SecondaryButton>test button</SecondaryButton>);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('renders with correct colors', () => {
      const { container } = render(<SecondaryButton color="blue">test button</SecondaryButton>);

      expect(container.firstChild).toHaveClass('text-blue-200', 'bg-blue-900', 'hover:bg-blue-800');
    });

    test('renders proper rounded styles', () => {
      const { container } = render(<SecondaryButton rounded>test button</SecondaryButton>);

      expect(container.firstChild).toHaveClass('rounded-full');
    });

    test('renders right icon with the correct classes', () => {
      const { container } = render(
        <SecondaryButton icon={PlusIcon} iconPosition="right">
          test button
        </SecondaryButton>
      );

      expect(container.firstChild).toHaveClass('inline-flex', 'flex-row-reverse');

      expect(container.querySelector('svg')).toHaveClass('ml-2', '-mr-1', 'h-5', 'w-5');
    });

    test('renders custom styles', () => {
      const { container } = render(
        <SecondaryButton
          className="text-green-700 bg-green-100 hover:bg-green-200 focus:ring-green-500"
          color="none"
        >
          test button
        </SecondaryButton>
      );

      expect(container.firstChild).toHaveClass(
        'text-green-700',
        'bg-green-100',
        'hover:bg-green-200',
        'focus:ring-green-500'
      );
    });
  });

  describe('OutlineButton', () => {
    test('renders correctly', () => {
      const { container } = render(<OutlineButton>test button</OutlineButton>);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('renders with correct colors', () => {
      const { container } = render(<OutlineButton color="blue">test button</OutlineButton>);

      expect(container.firstChild).toHaveClass(
        'text-blue-500',
        'border-blue-500',
        'hover:bg-blue-900'
      );
    });

    test('renders proper rounded styles', () => {
      const { container } = render(<OutlineButton rounded>test button</OutlineButton>);

      expect(container.firstChild).toHaveClass('rounded-full');
    });

    test('renders with correct sizes', () => {
      const { container } = render(<OutlineButton size="lg">test button</OutlineButton>);

      expect(container.firstChild).toHaveClass('px-4', 'py-2', 'text-base');
    });

    test('renders right icon with the correct classes', () => {
      const { container } = render(
        <OutlineButton icon={PlusIcon} iconPosition="right">
          test button
        </OutlineButton>
      );

      expect(container.firstChild).toHaveClass('inline-flex', 'flex-row-reverse');

      expect(container.querySelector('svg')).toHaveClass('ml-2', '-mr-1', 'h-5', 'w-5');
    });

    test('renders custom styles', () => {
      const { container } = render(
        <OutlineButton className="text-green-500 border-green-500 hover:bg-green-100" color="none">
          test button
        </OutlineButton>
      );

      expect(container.firstChild).toHaveClass(
        'text-green-500',
        'border-green-500',
        'hover:bg-green-100'
      );
    });
  });
});
