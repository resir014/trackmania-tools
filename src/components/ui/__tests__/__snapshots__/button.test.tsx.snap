// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button OutlineButton renders correctly 1`] = `
<button
  class="inline-flex flex-row px-4 py-2 text-sm rounded-md items-center justify-center border font-medium shadow-sm focus:outline-none focus:ring-2 focus:ring-green-500 focus:ring-offset-2 ring-offset-gray-900 text-blue-500 border-blue-500 hover:bg-blue-900 disabled:cursor-not-allowed disabled:opacity-75"
  type="button"
>
  <span>
    test button
  </span>
</button>
`;

exports[`Button PrimaryButton renders correctly 1`] = `
<button
  class="inline-flex flex-row px-4 py-2 text-sm rounded-md items-center justify-center border border-transparent font-medium shadow-sm focus:outline-none focus:ring-2 focus:ring-green-500 focus:ring-offset-2 ring-offset-gray-900 text-white bg-blue-600 hover:bg-blue-700 disabled:cursor-not-allowed disabled:opacity-75"
  type="button"
>
  <span>
    test button
  </span>
</button>
`;

exports[`Button SecondaryButton renders correctly 1`] = `
<button
  class="inline-flex flex-row px-4 py-2 text-sm rounded-md items-center justify-center border border-transparent font-medium focus:outline-none focus:ring-2 focus:ring-green-500 focus:ring-offset-2 ring-offset-gray-900 text-blue-200 bg-blue-900 hover:bg-blue-800 disabled:cursor-not-allowed disabled:opacity-75"
  type="button"
>
  <span>
    test button
  </span>
</button>
`;
